Toshi/Source/Toshi/Math/TQuaternion.cpp

```diff
@@ -1,142 +1,145 @@
+	void TQuaternion::SetIdentity()
+	{
+		Set(TQuaternion::IDENTITY.x, TQuaternion::IDENTITY.y, TQuaternion::IDENTITY.z, TQuaternion::IDENTITY.w);
+	}
 
-	if (mag > TMath::TFLOAT_EPSILON)
+	void TQuaternion::SetFromEulerYX(const float* fVal)
 	{
-		vec3.Divide(mag);
-		float dotProduct = TVector3::DotProduct(a_rVec3_2, a_rVec3);
-		float rot = 0.0f;
-		if ((dotProduct <= 1.0f - TMath::TFLOAT_EPSILON) && (rot = TMath::PI, -1.0f + TMath::TFLOAT_EPSILON <= dotProduct))
-		{
-			rot = TMath::ACos(dotProduct);
-		}
-		SetRotation(vec3, rot);
-		return;
+		float f1;
+		float f2;
+
+		TMath::SinCos(fVal[0] * 0.5f, f1, f2);
+		TQuaternion quat = TQuaternion(f1, 0, 0, f2);
+		TMath::SinCos(fVal[1] * 0.5f, f1, f2);
+		Set(0, f1, 0, f2);
+		*this *= quat;
 	}
 
-	float dotProduct = TVector3::DotProduct(a_rVec3_2, a_rVec3);
+	void TQuaternion::SetFromEulerRollPitchYaw(float a_fRoll, float a_fPitch, float a_fYaw)
+	{
+		float fCosRoll = cos(a_fRoll * 0.5f);
+		float fSinRoll = sin(a_fRoll * 0.5f);
+		float fCosPitch = cos(a_fPitch * 0.5f);
+		float fSinPitch = sin(a_fPitch * 0.5f);
+		float fCosYaw = cos(a_fYaw * 0.5f);
+		float fSinYaw = sin(a_fYaw * 0.5f);
+
+		float fX = fCosYaw * fCosPitch * fSinRoll - fSinYaw * fSinPitch * fCosRoll;
+		float fY = fCosPitch * fSinRoll * fSinYaw + fSinPitch * fCosRoll * fCosYaw;
+		float fZ = fCosRoll * fCosPitch * fSinYaw - fCosYaw * fSinPitch * fSinRoll;
+		float fW = fSinYaw * fSinPitch * fSinRoll + fCosYaw * fCosPitch * fCosRoll;
+
+		Set(fX, fY, fZ, fW);
+	}
+
+	void TQuaternion::SetRotation(const TVector3& a_rVec3, float a_fVal)
+	{
+		float fVal;
+		float fVal2;
+
+		TMath::SinCos(a_fVal * 0.5f, fVal, fVal2);
+		Set(a_rVec3.x * fVal, a_rVec3.y * fVal, a_rVec3.z * fVal, fVal2);
+	}
 
-	if (dotProduct < 0.0f)
+	void TQuaternion::SetVectorDirection(const TVector3& a_rVec3, const TVector3& a_rVec3_2)
 	{
-		TVector3 vec = a_rVec3_2;
-		if (TMath::Abs(a_rVec3_2.y) < 0.95f)
+		TVector3 vec3;
+		vec3.CrossProduct(a_rVec3_2, a_rVec3);
+
+		float mag = vec3.Magnitude();
+
+		if (mag > TMath::TFLOAT_EPSILON)
 		{
-			vec.RotateY(TMath::PI * 0.0055555557 * 90.0f);
+			vec3.Divide(mag);
+			float dotProduct = TVector3::DotProduct(a_rVec3_2, a_rVec3);
+			float rot = 0.0f;
+			if ((dotProduct <= 1.0f - TMath::TFLOAT_EPSILON) && (rot = TMath::PI, -1.0f + TMath::TFLOAT_EPSILON <= dotProduct))
+			{
+				rot = TMath::ACos(dotProduct);
+			}
+			SetRotation(vec3, rot);
+			return;
 		}
-		else
+
+		float dotProduct = TVector3::DotProduct(a_rVec3_2, a_rVec3);
+
+		if (dotProduct < 0.0f)
 		{
-			vec.RotateX(TMath::PI * 0.0055555557 * 90.0f);
+			TVector3 vec = a_rVec3_2;
+			if (TMath::Abs(a_rVec3_2.y) < 0.95f)
+			{
+				vec.RotateY(TMath::PI * 0.0055555557 * 90.0f);
+			}
+			else
+			{
+				vec.RotateX(TMath::PI * 0.0055555557 * 90.0f);
+			}
+			SetIdentity();
+			RotateAroundAxis(vec, TMath::PI * 0.0055555557 * 180.0f);
 		}
-		SetIdentity();
-		RotateAroundAxis(vec, TMath::PI * 0.0055555557 * 180.0f);
 	}
```

Comment:
Fuck i forgot, there needs to be another SetIdentity() in case the if statement is not met.

Reply:
I didn't really touch those functions, just moved them to the Toshi namespace